From 9c83cf1ed9bbea3b498cbd09e317996f9482f285 Mon Sep 17 00:00:00 2001
From: Matthias Brugger <mbrugger@suse.com>
Date: Thu, 20 Apr 2017 16:03:12 +0200
Subject: [PATCH 2/2] hns: set relation between hns_roce and hns_enet_drv
References: bsc#1033895
Patch-mainline: never, this is a temporary fix of the probe order

hns_roce relies on loading hns_enet_drv first. We hack a fixed call relation
which will fix module load and unload.

Signed-off-by: Matthias Brugger <mbrugger@suse.com>
---
 drivers/infiniband/hw/hns/hns_roce_main.c     |  5 +++++
 drivers/net/ethernet/hisilicon/hns/hns_enet.c | 12 ++++++++++++
 include/linux/hns/hns.h                       |  7 +++++++
 3 files changed, 24 insertions(+)
 create mode 100644 include/linux/hns/hns.h

diff --git a/drivers/infiniband/hw/hns/hns_roce_main.c b/drivers/infiniband/hw/hns/hns_roce_main.c
index 8acb290de68c..b70531dd492a 100644
--- a/drivers/infiniband/hw/hns/hns_roce_main.c
+++ b/drivers/infiniband/hw/hns/hns_roce_main.c
@@ -41,6 +41,7 @@
 #include "hns_roce_device.h"
 #include "hns_roce_user.h"
 #include "hns_roce_hem.h"
+#include <linux/hns/hns.h>
 
 /**
  * hns_get_gid_index - Get gid index.
@@ -841,6 +842,8 @@ static int hns_roce_probe(struct platform_device *pdev)
 	struct hns_roce_dev *hr_dev;
 	struct device *dev = &pdev->dev;
 
+	hns_enet_register();
+
 	hr_dev = (struct hns_roce_dev *)ib_alloc_device(sizeof(*hr_dev));
 	if (!hr_dev)
 		return -ENOMEM;
@@ -965,6 +968,8 @@ static int hns_roce_remove(struct platform_device *pdev)
 
 	ib_dealloc_device(&hr_dev->ib_dev);
 
+	hns_enet_unregister();
+
 	return 0;
 }
 
diff --git a/drivers/net/ethernet/hisilicon/hns/hns_enet.c b/drivers/net/ethernet/hisilicon/hns/hns_enet.c
index c6700b91a2df..a1e927bb2a3e 100644
--- a/drivers/net/ethernet/hisilicon/hns/hns_enet.c
+++ b/drivers/net/ethernet/hisilicon/hns/hns_enet.c
@@ -2388,6 +2388,18 @@ static int hns_nic_dev_remove(struct platform_device *pdev)
 	return 0;
 }
 
+void hns_enet_register(void)
+{
+	__module_get(THIS_MODULE);
+}
+EXPORT_SYMBOL(hns_enet_register);
+
+void hns_enet_unregister(void)
+{
+	module_put(THIS_MODULE);
+}
+EXPORT_SYMBOL(hns_enet_unregister);
+
 static const struct of_device_id hns_enet_of_match[] = {
 	{.compatible = "hisilicon,hns-nic-v1",},
 	{.compatible = "hisilicon,hns-nic-v2",},
diff --git a/include/linux/hns/hns.h b/include/linux/hns/hns.h
new file mode 100644
index 000000000000..a9371b6f3d64
--- /dev/null
+++ b/include/linux/hns/hns.h
@@ -0,0 +1,7 @@
+#ifndef __HNS_MDIO_H
+#define __HNS_MDIO_H
+
+void hns_enet_register(void);
+void hns_enet_unregister(void);
+
+#endif
-- 
2.12.0

